libvips/foreign/webpsave.c

```diff
@@ -109,15 +155,639 @@ typedef struct _VipsForeignSaveWebp {
 
 	/* Profile to embed.
 	 */
-	char *profile;
+	const char *profile;
+
+	WebPConfig config;
+
+	/* Output is written here. We can only support memory write, since we
+	 * handle metadata.
+	 */
+	WebPMemoryWriter memory_writer;
+
+	/* Write animated webp here.
+	*/
+	WebPAnimEncoder *enc;
+
+	/* Add metadata with this.
+	 */
+	WebPMux *mux;
+
+	/* The current frame coming from libvips, and the y position
+	 * in the input image.
+	 */
+	VipsRegion *frame;
+	int write_y;
 
-} VipsForeignSaveWebp;
+	/* VipsRegion is not always contiguious, but we need contiguous RGB(A)
+	 * for libwebp. We need to copy each frame to a local buffer.
+	 */
+	VipsPel *frame_bytes;
+} VipsForeignSaveWebP;
 
-typedef VipsForeignSaveClass VipsForeignSaveWebpClass;
+typedef VipsForeignSaveClass VipsForeignSaveWebPClass;
 
-G_DEFINE_ABSTRACT_TYPE( VipsForeignSaveWebp, vips_foreign_save_webp, 
+G_DEFINE_ABSTRACT_TYPE( VipsForeignSaveWebP, vips_foreign_save_webp,
 	VIPS_TYPE_FOREIGN_SAVE );
 
+static void
+vips_webp_write_unset( VipsForeignSaveWebP *write )
+{
+	WebPMemoryWriterClear( &write->memory_writer );
+	VIPS_FREEF( WebPAnimEncoderDelete, write->enc );
+	VIPS_FREEF( WebPMuxDelete, write->mux );
+	VIPS_UNREF( write->image );
+}
+
+static void
+vips_foreign_save_webp_dispose( GObject *gobject )
+{
+	VipsForeignSaveWebP *webp= (VipsForeignSaveWebP *) gobject;
+
+	VIPS_UNREF( webp->frame );
+
+	VIPS_UNREF( webp->target );
+
+	VIPS_FREE( webp->frame_bytes );
+
+	G_OBJECT_CLASS( vips_foreign_save_webp_parent_class )->
+		dispose( gobject );
+}
+
+static gboolean
+vips_webp_pic_init( VipsForeignSaveWebP *write, WebPPicture *pic )
+{
+	if( !WebPPictureInit( pic ) ) {
+		vips_error( "webpsave", "%s", _( "picture version error" ) );
+		return( FALSE );
+	}
+	pic->writer = WebPMemoryWrite;
+	pic->custom_ptr = (void *) &write->memory_writer;
+
+	/* Smart subsampling needs use_argb because it is applied during
+	 * RGB to YUV conversion.
+	 */
+	pic->use_argb = write->lossless ||
+		write->near_lossless ||
+		write->smart_subsample;
+
+	return( TRUE );
+}
+
+/* Write a VipsImage into an unintialised pic.
+ */
+static int
+write_webp_image( VipsForeignSaveWebP *write, const VipsPel *imagedata, WebPPicture *pic )
```

Comment:
`vips_foreign_save_webp_write_webp_image()` maybe?

The line needs to fit in 80 columns, so maybe wrap before `const VipsPel ..`.